aptos-move/framework/aptos-token/sources/bid.move

```diff
@@ -0,0 +1,343 @@
+
+        if (!exists<BidRecords<CoinType>>(owner_addr)) {
+            move_to(
+                bidder,
+                BidRecords<CoinType> {
+                    records: table::new(),
+                    bid_event: account::new_event_handle<BidEvent<CoinType>>(bidder),
+                    withdraw_bid_event: account::new_event_handle<WithdrawBidEvent<CoinType>>(bidder),
+                }
+            );
+        };
+    }
+
+    public fun assert_bid_for_a_listing<CoinType>(
+        token_id: TokenId,
+        offer_price: u64,
+        token_amount: u64,
+        entry: &Listing<CoinType>
+    ) {
+        // validate token_id match
+        assert!(token_id == listing::get_listing_token_id(entry), ETOKEN_ID_NOT_MATCH);
+        // validate offerred amount and price
+        let listed_amount =  listing::get_listing_token_amount(entry);
+        let min_total = listing::get_listing_min_price(entry) * listed_amount;
+        assert!(offer_price * token_amount >= min_total, ENO_SUFFICIENT_FUND);
+        assert!(token_amount == listed_amount, ETOKEN_AMOUNT_NOT_MATCH);
+    }
+
+    /// withdraw the coin and store them in bid struct and return a global unique bid id
+    public fun bid<CoinType>(
+        bidder: &signer,
+        token_id: TokenId,
+        token_amount:u64,
+        offer_price: u64,
+        entry: &Listing<CoinType>,
+    ): ID acquires BidRecords {
+        initialize_bid_records<CoinType>(bidder);
+        let bidder_address = signer::address_of(bidder);
+        // check the bid is legit for the listing
+        assert_bid_for_a_listing(token_id, offer_price * token_amount, token_amount, entry);
+        // check bidder has sufficient balance
+        assert!(coin::balance<CoinType>(bidder_address) >= offer_price * token_amount, error::invalid_argument(ENO_SUFFICIENT_FUND));
+        // withdraw the coin and store them in escrow to ensure the fund is avaliable until expiration_sec
+        let coin = coin::withdraw<CoinType>(bidder, offer_price * token_amount);
+
+        let bid = Bid<CoinType> {
+            bidder: bidder_address,
+            coin,
+            offer_price,
+            listing_id: listing::get_listing_id(entry),
+        };
+        let bid_id = create_bid_id(bidder);
+        initialize_bid_records<CoinType>(bidder);
+        let bid_records = borrow_global_mut<BidRecords<CoinType>>(bidder_address);
+        table::add(&mut bid_records.records, bid_id, bid);
+        event::emit_event<BidEvent<CoinType>>(
+            &mut bid_records.bid_event,
+            BidEvent<CoinType> {
+                offer_price,
+                listing_id: listing::get_listing_id(entry),
+                bid_id,
+            },
+        );
+
+        bid_id
+    }
+
+    /// bidder can withdraw fund from any bid when lock time ends
+    public entry fun withdraw_fund_from_bid<CoinType>(
+        bidder: &signer,
+        bid_id_creation_number: u64,
+    ) acquires BidRecords {
```

Comment:
naming feels like it should be more like `cancel_bid` or just `withdraw_bid`.

Reply:
cancel_bid is defined in the auction_house, where the auction owner removes one bid from `bids` in `Auction`.  This removed bid won't participate in the bidding.

withdraw_fund_from_bid here is for the bidder to extract the coins from his bid without any approval from the Auction House

Reply:
I don't know what the purpose of this is but to effectively cancel my own bid. withdraw_fund(s) from bid doesn't necessarily imply you would remove the bid from the bidding table.

Reply:
Yes, withdraw doesn't remove bid from listing. The bid library knows nothing about the auction house's data. 
The purpose is to allow bidder to extract their coin and store into their account after the auction ends without the auctioneer's approval.  renamed